text/0000-ember-data-return-promise-from-ds-model-save.md

```diff
@@ -0,0 +1,65 @@
+* Start Date: 08/22/2018
+* RFC PR: (leave this empty)
+* Ember Issue: (leave this empty)
+
+# Return a Promise from DS.Model.save()
+
+## Summary
+
+DS.Model.save() will return an RSVP.Promise instead of a PromiseObject.
+
+## Motivation
+
+* The API documentation already documents the return value as a Promise.
+* Remove dependency on promise proxies
+* Async Consistency - The PromiseObject encourages a usage pattern of sometimes-async, sometimes-sync behavior. We want to [reduce zalgo](http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony).
+* API consistency - While the PromiseObject proxies properties, the PromiseObject does not proxy methods on the underlying model. For instance, if you try to call a method like destroyRecord on the model, you will get a “not a function” error because the method is called on the Proxy and not the underlying model. This encourages users for Ember Data to reach into private implementation details of the promise proxy, such as using proxy.get('content') to call methods.
+* Enable new functionality in Ember Data while making backwards compatibility with older versions of Ember easier.
+* The documentation of using PromiseProxy APIs has been replaced over time in the guides and API documentation to use more typical JavaScript usage of Promises, such as async/await, instead of PromiseProxies.
+
+## Detailed design
+
+Introduce a new feature flag, “ds-return-promise-from-model-save”. When enabled, DS.Model.save() will return an RSVP.Promise that resolves with the model if the save was successful, or rejects with an error if the save fails. When disabled, a PromiseObject will be returned to keep today's behavior, but accessing properties on the PromiseObject or calling non-Promise functions (.then, .finally, .catch) on the PromiseObject will issue a deprecation warning.
```

Comment:
We should go directly from canary => beta => release with the deprecation on. I don't believe we need or should use a feature flag here to guard the new world, as promiseObjects do function as normal promises if used as such.

That said, @rwjblue may have some thoughts, and using the flag as suggested below would let us code-strip for users that opt-in to a post-proxy world.

Reply:
I take this back and we should use a feature flag.